File: Source/Nava.Runtime.CS/Disposable.cs
namespace Nava.Runtime;
/// <summary> Abstract base class for disposable types. </summary>
public abstract class Disposable : IDisposable {
	/// <summary> Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources. </summary>
	/// <param name="disposing"><c>true</c> if called from <see cref="Disposable.Dispose()"/>.</param>
	protected abstract void Dispose(bool disposing);

	/// <summary> Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources. </summary>
	public void Dispose() {
		GC.SuppressFinalize(this);
		Dispose(true);
	}
	/// <summary> Finalizes an instance of the <see cref="Disposable"/> class. </summary>
	~Disposable() => Dispose(false);
}

File: Source/Nava.Runtime/Disposable.fs
namespace Nava.Runtime

open System

[<AbstractClass>]
type DisposableFS =
    new() = { }

    /// <summary> Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources. </summary>
    /// <param name="disposing"><c>true</c> if called from <see cref="M:Nava.Runtime.Disposable.Dispose" />.</param>
    abstract member Dispose: disposing: bool -> unit

    /// <summary> Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources. </summary>
    member this.Dispose() =
        GC.SuppressFinalize this
        this.Dispose true

    override this.Finalize() = this.Dispose false

    interface IDisposable with
        member this.Dispose() = this.Dispose()